<a href="https://colab.research.google.com/github/Hashim28/Guvi-task-/blob/main/program.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [204]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime, timedelta

# **1.Users Data**

In [205]:
users = pd.read_csv('https://raw.githubusercontent.com/Hashim28/Guvi-task-/main/takehome_users.csv',encoding='latin-1')

In [206]:
users

object_id        creation_time               name  \
0              1  2014-04-22 03:53:30     Clausen August   
1              2  2013-11-15 03:45:04      Poole Matthew   
2              3  2013-03-19 23:14:52  Bottrill Mitchell   
3              4  2013-05-21 08:09:28    Clausen Nicklas   
4              5  2013-01-17 10:14:20          Raw Grace   
...          ...                  ...                ...   
11995      11996  2013-09-06 06:14:15       Meier Sophia   
11996      11997  2013-01-10 18:28:37      Fisher Amelie   
11997      11998  2014-04-27 12:45:16        Haynes Jake   
11998      11999  2012-05-31 11:55:59       Faber Annett   
11999      12000  2014-01-26 08:57:12         Lima Thaís   

                            email     creation_source  \
0        AugustCClausen@yahoo.com        GUEST_INVITE   
1          MatthewPoole@gustr.com          ORG_INVITE   
2      MitchellBottrill@gustr.com          ORG_INVITE   
3       NicklasSClausen@yahoo.com        GUEST_INVITE   
4              GraceRaw@yahoo.com        GUEST_INVITE   
...                           ...                 ...   
11995       SophiaMeier@gustr.com          ORG_INVITE   
11996      AmelieFisher@gmail.com  SIGNUP_GOOGLE_AUTH   
11997         JakeHaynes@cuvox.de        GUEST_INVITE   
11998          mhaerzxp@iuxiw.com   PERSONAL_PROJECTS   
11999   ThaisMeloLima@hotmail.com              SIGNUP   

       last_session_creation_time  opted_in_to_mailing_list  \
0                    1.398139e+09                         1   
1                    1.396238e+09                         0   
2                    1.363735e+09                         0   
3                    1.369210e+09                         0   
4                    1.358850e+09                         0   
...                           ...                       ...   
11995                1.378448e+09                         0   
11996                1.358275e+09                         0   
11997                1.398603e+09                         1   
11998                1.338638e+09                         0   
11999                1.390727e+09                         0   

       enabled_for_marketing_drip  org_id  invited_by_user_id  
0                               0      11             10803.0  
1                               0       1               316.0  
2                               0      94              1525.0  
3                               0       1              5151.0  
4                               0     193              5240.0  
...                           ...     ...                 ...  
11995                           0      89              8263.0  
11996                           0     200                 NaN  
11997                           1      83              8074.0  
11998                           0       6                 NaN  
11999                           1       0                 NaN  

[12000 rows x 10 columns]

In [207]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12000 entries, 0 to 11999
Data columns (total 10 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   object_id                   12000 non-null  int64  
 1   creation_time               12000 non-null  object 
 2   name                        12000 non-null  object 
 3   email                       12000 non-null  object 
 4   creation_source             12000 non-null  object 
 5   last_session_creation_time  8823 non-null   float64
 6   opted_in_to_mailing_list    12000 non-null  int64  
 7   enabled_for_marketing_drip  12000 non-null  int64  
 8   org_id                      12000 non-null  int64  
 9   invited_by_user_id          6417 non-null   float64
dtypes: float64(2), int64(4), object(4)
memory usage: 937.6+ KB


In [208]:
users.describe().T

count          mean           std           min  \
object_id                   12000.0  6.000500e+03  3.464246e+03  1.000000e+00   
last_session_creation_time   8823.0  1.379279e+09  1.953116e+07  1.338452e+09   
opted_in_to_mailing_list    12000.0  2.495000e-01  4.327417e-01  0.000000e+00   
enabled_for_marketing_drip  12000.0  1.493333e-01  3.564316e-01  0.000000e+00   
org_id                      12000.0  1.418846e+02  1.240567e+02  0.000000e+00   
invited_by_user_id           6417.0  5.962957e+03  3.383762e+03  3.000000e+00   

                                     25%           50%           75%  \
object_id                   3.000750e+03  6.000500e+03  9.000250e+03   
last_session_creation_time  1.363195e+09  1.382888e+09  1.398443e+09   
opted_in_to_mailing_list    0.000000e+00  0.000000e+00  0.000000e+00   
enabled_for_marketing_drip  0.000000e+00  0.000000e+00  0.000000e+00   
org_id                      2.900000e+01  1.080000e+02  2.382500e+02   
invited_by_user_id          3.058000e+03  5.954000e+03  8.817000e+03   

                                     max  
object_id                   1.200000e+04  
last_session_creation_time  1.402067e+09  
opted_in_to_mailing_list    1.000000e+00  
enabled_for_marketing_drip  1.000000e+00  
org_id                      4.160000e+02  
invited_by_user_id          1.199900e+04

# **2.Users Engagement data**

In [209]:
users_engagement = pd.read_csv('https://raw.githubusercontent.com/Hashim28/Guvi-task-/main/takehome_user_engagement.csv')

In [210]:
users_engagement

time_stamp  user_id  visited
0       2014-04-22 03:53:30        1        1
1       2013-11-15 03:45:04        2        1
2       2013-11-29 03:45:04        2        1
3       2013-12-09 03:45:04        2        1
4       2013-12-25 03:45:04        2        1
...                     ...      ...      ...
207912  2013-09-06 06:14:15    11996        1
207913  2013-01-15 18:28:37    11997        1
207914  2014-04-27 12:45:16    11998        1
207915  2012-06-02 11:55:59    11999        1
207916  2014-01-26 08:57:12    12000        1

[207917 rows x 3 columns]

In [211]:
users_engagement.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207917 entries, 0 to 207916
Data columns (total 3 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   time_stamp  207917 non-null  object
 1   user_id     207917 non-null  int64 
 2   visited     207917 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 4.8+ MB


In [212]:
users_engagement.describe().T

count         mean          std  min     25%     50%     75%  \
user_id  207917.0  5913.314197  3394.941674  1.0  3087.0  5682.0  8944.0   
visited  207917.0     1.000000     0.000000  1.0     1.0     1.0     1.0   

             max  
user_id  12000.0  
visited      1.0

# Checking for null values in both data

In [213]:
users.isna().sum()

object_id                        0
creation_time                    0
name                             0
email                            0
creation_source                  0
last_session_creation_time    3177
opted_in_to_mailing_list         0
enabled_for_marketing_drip       0
org_id                           0
invited_by_user_id            5583
dtype: int64

In [214]:
users_engagement.isna().sum()

time_stamp    0
user_id       0
visited       0
dtype: int64

users dataframe has some null values in the _last_session_creation_time and _invited_by_userid columns

# Checking for duplicates

In [215]:
users.duplicated().value_counts()

False    12000
dtype: int64

In [216]:
users_engagement.duplicated().value_counts()

False    207917
dtype: int64

No duplcates in both data

### Converting creation_time and time_stamp to datetime datatype

In [217]:
users['creation_time'] = pd.to_datetime(users['creation_time'])

In [218]:
users['creation_time']

0       2014-04-22 03:53:30
1       2013-11-15 03:45:04
2       2013-03-19 23:14:52
3       2013-05-21 08:09:28
4       2013-01-17 10:14:20
                ...        
11995   2013-09-06 06:14:15
11996   2013-01-10 18:28:37
11997   2014-04-27 12:45:16
11998   2012-05-31 11:55:59
11999   2014-01-26 08:57:12
Name: creation_time, Length: 12000, dtype: datetime64[ns]

In [219]:
users_engagement['time_stamp'] = pd.to_datetime(users_engagement['time_stamp'])

In [220]:
users_engagement['time_stamp'] 

0        2014-04-22 03:53:30
1        2013-11-15 03:45:04
2        2013-11-29 03:45:04
3        2013-12-09 03:45:04
4        2013-12-25 03:45:04
                 ...        
207912   2013-09-06 06:14:15
207913   2013-01-15 18:28:37
207914   2014-04-27 12:45:16
207915   2012-06-02 11:55:59
207916   2014-01-26 08:57:12
Name: time_stamp, Length: 207917, dtype: datetime64[ns]

In [221]:
users['last_session_creation_time'] = pd.to_datetime(users['last_session_creation_time'],unit = 's')

In [222]:
users['last_session_creation_time']

0       2014-04-22 03:53:30
1       2014-03-31 03:45:04
2       2013-03-19 23:14:52
3       2013-05-22 08:09:28
4       2013-01-22 10:14:20
                ...        
11995   2013-09-06 06:14:15
11996   2013-01-15 18:28:37
11997   2014-04-27 12:45:16
11998   2012-06-02 11:55:59
11999   2014-01-26 08:57:12
Name: last_session_creation_time, Length: 12000, dtype: datetime64[ns]

In [223]:
users['object_id'].nunique()

12000

In [224]:
users_engagement['user_id'].nunique()

8823

In Users engagement dataframe, there is data for 8823 users out of 12000 users

# Groupby and rolling methods can be used to find the users who were active for more than 3 days in any of the 7 day rolling period.

In [225]:
df = users_engagement.copy()

In [226]:
df

time_stamp  user_id  visited
0      2014-04-22 03:53:30        1        1
1      2013-11-15 03:45:04        2        1
2      2013-11-29 03:45:04        2        1
3      2013-12-09 03:45:04        2        1
4      2013-12-25 03:45:04        2        1
...                    ...      ...      ...
207912 2013-09-06 06:14:15    11996        1
207913 2013-01-15 18:28:37    11997        1
207914 2014-04-27 12:45:16    11998        1
207915 2012-06-02 11:55:59    11999        1
207916 2014-01-26 08:57:12    12000        1

[207917 rows x 3 columns]

In [227]:
df['visited_7_days'] = df.groupby('user_id',as_index = False,group_keys = False).apply(lambda x:x.rolling('7D',on = 'time_stamp')['user_id'].count())

In [228]:
df

time_stamp  user_id  visited  visited_7_days
0      2014-04-22 03:53:30        1        1             1.0
1      2013-11-15 03:45:04        2        1             1.0
2      2013-11-29 03:45:04        2        1             1.0
3      2013-12-09 03:45:04        2        1             1.0
4      2013-12-25 03:45:04        2        1             1.0
...                    ...      ...      ...             ...
207912 2013-09-06 06:14:15    11996        1             1.0
207913 2013-01-15 18:28:37    11997        1             1.0
207914 2014-04-27 12:45:16    11998        1             1.0
207915 2012-06-02 11:55:59    11999        1             1.0
207916 2014-01-26 08:57:12    12000        1             1.0

[207917 rows x 4 columns]

If any of the values in visited_7_days column is greater than 2, then it is saved in adopted user column

In [229]:
user_adopted = df.groupby('user_id')['visited_7_days'].max().reset_index()

In [230]:
user_adopted.head()

user_id  visited_7_days
0        1             1.0
1        2             3.0
2        3             1.0
3        4             1.0
4        5             1.0

In [231]:
user_adopted['adopted_user'] = user_adopted['visited_7_days'].apply(lambda x: 1 if x >= 3 else 0)

In [232]:
user_adopted.head()

user_id  visited_7_days  adopted_user
0        1             1.0             0
1        2             3.0             1
2        3             1.0             0
3        4             1.0             0
4        5             1.0             0

In [233]:
user_adopted['adopted_user'].value_counts()

0    7221
1    1602
Name: adopted_user, dtype: int64

# Joining user_adopted to users dataframe

In [234]:
user_adopted.drop('visited_7_days',axis = 1, inplace = True)

In [235]:
user_adopted.rename(columns = {'user_id':'object_id'},inplace = True)

In [236]:
user_adopted.set_index('object_id',inplace = True)

In [237]:
user_adopted

adopted_user
object_id              
1                     0
2                     1
3                     0
4                     0
5                     0
...                 ...
11996                 0
11997                 0
11998                 0
11999                 0
12000                 0

[8823 rows x 1 columns]

In [238]:
df_users = users.join(user_adopted, on = 'object_id', how = 'left')

In [239]:
df_users

object_id       creation_time               name  \
0              1 2014-04-22 03:53:30     Clausen August   
1              2 2013-11-15 03:45:04      Poole Matthew   
2              3 2013-03-19 23:14:52  Bottrill Mitchell   
3              4 2013-05-21 08:09:28    Clausen Nicklas   
4              5 2013-01-17 10:14:20          Raw Grace   
...          ...                 ...                ...   
11995      11996 2013-09-06 06:14:15       Meier Sophia   
11996      11997 2013-01-10 18:28:37      Fisher Amelie   
11997      11998 2014-04-27 12:45:16        Haynes Jake   
11998      11999 2012-05-31 11:55:59       Faber Annett   
11999      12000 2014-01-26 08:57:12         Lima Thaís   

                            email     creation_source  \
0        AugustCClausen@yahoo.com        GUEST_INVITE   
1          MatthewPoole@gustr.com          ORG_INVITE   
2      MitchellBottrill@gustr.com          ORG_INVITE   
3       NicklasSClausen@yahoo.com        GUEST_INVITE   
4              GraceRaw@yahoo.com        GUEST_INVITE   
...                           ...                 ...   
11995       SophiaMeier@gustr.com          ORG_INVITE   
11996      AmelieFisher@gmail.com  SIGNUP_GOOGLE_AUTH   
11997         JakeHaynes@cuvox.de        GUEST_INVITE   
11998          mhaerzxp@iuxiw.com   PERSONAL_PROJECTS   
11999   ThaisMeloLima@hotmail.com              SIGNUP   

      last_session_creation_time  opted_in_to_mailing_list  \
0            2014-04-22 03:53:30                         1   
1            2014-03-31 03:45:04                         0   
2            2013-03-19 23:14:52                         0   
3            2013-05-22 08:09:28                         0   
4            2013-01-22 10:14:20                         0   
...                          ...                       ...   
11995        2013-09-06 06:14:15                         0   
11996        2013-01-15 18:28:37                         0   
11997        2014-04-27 12:45:16                         1   
11998        2012-06-02 11:55:59                         0   
11999        2014-01-26 08:57:12                         0   

       enabled_for_marketing_drip  org_id  invited_by_user_id  adopted_user  
0                               0      11             10803.0           0.0  
1                               0       1               316.0           1.0  
2                               0      94              1525.0           0.0  
3                               0       1              5151.0           0.0  
4                               0     193              5240.0           0.0  
...                           ...     ...                 ...           ...  
11995                           0      89              8263.0           0.0  
11996                           0     200                 NaN           0.0  
11997                           1      83              8074.0           0.0  
11998                           0       6                 NaN           0.0  
11999                           1       0                 NaN           0.0  

[12000 rows x 11 columns]

In [240]:
df_users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12000 entries, 0 to 11999
Data columns (total 11 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   object_id                   12000 non-null  int64         
 1   creation_time               12000 non-null  datetime64[ns]
 2   name                        12000 non-null  object        
 3   email                       12000 non-null  object        
 4   creation_source             12000 non-null  object        
 5   last_session_creation_time  8823 non-null   datetime64[ns]
 6   opted_in_to_mailing_list    12000 non-null  int64         
 7   enabled_for_marketing_drip  12000 non-null  int64         
 8   org_id                      12000 non-null  int64         
 9   invited_by_user_id          6417 non-null   float64       
 10  adopted_user                8823 non-null   float64       
dtypes: datetime64[ns](2), float64(2), int64(4), object(3)


The null values in the _adopted_user and _last_session_creation_time can be filled in with 0 by assuming that those users are not adopted users.

In [241]:
df_users['last_session_creation_time'].fillna(0, inplace = True)
df_users['adopted_user'].fillna(0, inplace = True)

Dropping columns which are not useful.

 last_session_creation_time and creation_time can be removed abecause it was used to create the _adopted_user column.

In [242]:
df_users.drop(['object_id','name','email'], axis = 1, inplace = True)

In [244]:
df_users.drop(['last_session_creation_time','creation_time'], axis = 1, inplace = True)

Using get_dummies()

In [246]:
df_users = pd.get_dummies(df_users, columns = ['creation_source'])

In [247]:
df_users

opted_in_to_mailing_list  enabled_for_marketing_drip  org_id  \
0                             1                           0      11   
1                             0                           0       1   
2                             0                           0      94   
3                             0                           0       1   
4                             0                           0     193   
...                         ...                         ...     ...   
11995                         0                           0      89   
11996                         0                           0     200   
11997                         1                           1      83   
11998                         0                           0       6   
11999                         0                           1       0   

       invited_by_user_id  adopted_user  creation_source_GUEST_INVITE  \
0                 10803.0           0.0                             1   
1                   316.0           1.0                             0   
2                  1525.0           0.0                             0   
3                  5151.0           0.0                             1   
4                  5240.0           0.0                             1   
...                   ...           ...                           ...   
11995              8263.0           0.0                             0   
11996                 NaN           0.0                             0   
11997              8074.0           0.0                             1   
11998                 NaN           0.0                             0   
11999                 NaN           0.0                             0   

       creation_source_ORG_INVITE  creation_source_PERSONAL_PROJECTS  \
0                               0                                  0   
1                               1                                  0   
2                               1                                  0   
3                               0                                  0   
4                               0                                  0   
...                           ...                                ...   
11995                           1                                  0   
11996                           0                                  0   
11997                           0                                  0   
11998                           0                                  1   
11999                           0                                  0   

       creation_source_SIGNUP  creation_source_SIGNUP_GOOGLE_AUTH  
0                           0                                   0  
1                           0                                   0  
2                           0                                   0  
3                           0                                   0  
4                           0                                   0  
...                       ...                                 ...  
11995                       0                                   0  
11996                       0                                   1  
11997                       0                                   0  
11998                       0                                   0  
11999                       1                                   0  

[12000 rows x 10 columns]

In [248]:
df_users.describe().T

count         mean          std  min  \
opted_in_to_mailing_list            12000.0     0.249500     0.432742  0.0   
enabled_for_marketing_drip          12000.0     0.149333     0.356432  0.0   
org_id                              12000.0   141.884583   124.056723  0.0   
invited_by_user_id                   6417.0  5962.957145  3383.761968  3.0   
adopted_user                        12000.0     0.133500     0.340128  0.0   
creation_source_GUEST_INVITE        12000.0     0.180250     0.384412  0.0   
creation_source_ORG_INVITE          12000.0     0.354500     0.478381  0.0   
creation_source_PERSONAL_PROJECTS   12000.0     0.175917     0.380765  0.0   
creation_source_SIGNUP              12000.0     0.173917     0.379054  0.0   
creation_source_SIGNUP_GOOGLE_AUTH  12000.0     0.115417     0.319537  0.0   

                                       25%     50%      75%      max  
opted_in_to_mailing_list               0.0     0.0     0.00      1.0  
enabled_for_marketing_drip             0.0     0.0     0.00      1.0  
org_id                                29.0   108.0   238.25    416.0  
invited_by_user_id                  3058.0  5954.0  8817.00  11999.0  
adopted_user                           0.0     0.0     0.00      1.0  
creation_source_GUEST_INVITE           0.0     0.0     0.00      1.0  
creation_source_ORG_INVITE             0.0     0.0     1.00      1.0  
creation_source_PERSONAL_PROJECTS      0.0     0.0     0.00      1.0  
creation_source_SIGNUP                 0.0     0.0     0.00      1.0  
creation_source_SIGNUP_GOOGLE_AUTH     0.0     0.0     0.00      1.0

In [252]:
df_users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12000 entries, 0 to 11999
Data columns (total 10 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   opted_in_to_mailing_list            12000 non-null  int64  
 1   enabled_for_marketing_drip          12000 non-null  int64  
 2   org_id                              12000 non-null  int64  
 3   invited_by_user_id                  6417 non-null   float64
 4   adopted_user                        12000 non-null  float64
 5   creation_source_GUEST_INVITE        12000 non-null  uint8  
 6   creation_source_ORG_INVITE          12000 non-null  uint8  
 7   creation_source_PERSONAL_PROJECTS   12000 non-null  uint8  
 8   creation_source_SIGNUP              12000 non-null  uint8  
 9   creation_source_SIGNUP_GOOGLE_AUTH  12000 non-null  uint8  
dtypes: float64(2), int64(3), uint8(5)
memory usage: 527.5 KB


In [251]:
df_users.corr()['adopted_user']

opted_in_to_mailing_list              0.008097
enabled_for_marketing_drip            0.004653
org_id                                0.064523
invited_by_user_id                    0.000675
adopted_user                          1.000000
creation_source_GUEST_INVITE          0.045408
creation_source_ORG_INVITE           -0.007636
creation_source_PERSONAL_PROJECTS    -0.075817
creation_source_SIGNUP                0.009299
creation_source_SIGNUP_GOOGLE_AUTH    0.036119
Name: adopted_user, dtype: float64